pkg/doublewrite/doublewritebackend.go

```diff
@@ -0,0 +1,198 @@
+	}
+	kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
+	}
+	if d.readFromKVStore {
+		return kvStoreErr
+	}
+	return crdErr
+}
+
+func (d *doubleWriteBackend) UpdateKeyIfLocked(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool, lock kvstore.KVLocker) error {
+	crdErr := d.crdBackend.UpdateKeyIfLocked(ctx, id, key, true, lock)
+	if crdErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key with lock")
+	}
+	kvStoreErr := d.kvstoreBackend.UpdateKeyIfLocked(ctx, id, key, true, lock)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key with lock")
+	}
+	if d.readFromKVStore {
+		return kvStoreErr
+	}
+	return crdErr
+}
+
+func (d *doubleWriteBackend) Lock(ctx context.Context, key allocator.AllocatorKey) (kvstore.KVLocker, error) {
+	// CRD backend doesn't require locking but since we are still doing KVStore operations, let's use the KVStore lock
+	return d.kvstoreBackend.Lock(ctx, key)
+}
+
+func (d *doubleWriteBackend) Get(ctx context.Context, key allocator.AllocatorKey) (idpool.ID, error) {
+	if d.readFromKVStore {
+		return d.kvstoreBackend.Get(ctx, key)
+	}
+	return d.crdBackend.Get(ctx, key)
+}
+
+func (d *doubleWriteBackend) GetIfLocked(ctx context.Context, key allocator.AllocatorKey, lock kvstore.KVLocker) (idpool.ID, error) {
+	if d.readFromKVStore {
+		return d.kvstoreBackend.GetIfLocked(ctx, key, lock)
+	}
+	return d.crdBackend.GetIfLocked(ctx, key, lock)
+}
+
+func (d *doubleWriteBackend) GetByID(ctx context.Context, id idpool.ID) (allocator.AllocatorKey, error) {
+	if d.readFromKVStore {
+		return d.kvstoreBackend.GetByID(ctx, id)
+	}
+	return d.crdBackend.GetByID(ctx, id)
+}
+
+func (d *doubleWriteBackend) Release(ctx context.Context, id idpool.ID, key allocator.AllocatorKey) (err error) {
+	kvStoreErr := d.kvstoreBackend.Release(ctx, id, key)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to release identity")
+	}
+	// This is a no-op for the CRD backend
+	if d.readFromKVStore {
+		return kvStoreErr
+	}
+	return nil
+}
+
+func (d *doubleWriteBackend) ListAndWatch(ctx context.Context, handler allocator.CacheMutations, stopChan chan struct{}) {
+	d.crdBackend.ListAndWatch(ctx, handler, stopChan)
+	d.kvstoreBackend.ListAndWatch(ctx, handler, stopChan)
```

Comment:
You are right, closing the channel was an issue, I refactored the code to account for that.
There is one subtlety: the `crdBackend` cannot do any read operations if `ListAndWatch` has not been called beforehand, because it's the only place where the underlying `Store` is initialized:
https://github.com/cilium/cilium/blob/797330cee40330c505fb97f6dd38446afb96ec78/pkg/k8s/identitybackend/identity.go#L369-L371
I didn't attempt to refactor that because it would be a non-trivial change. However, this forced me to call `ListAndWatch` in a number of places to force the initialization of the `Store`:
https://github.com/cilium/cilium/blob/f71bbf171602d800d63324db9d348717399c6b94/pkg/kvstore/doublewrite/backend.go#L198-L210
https://github.com/cilium/cilium/blob/240f36651260be0cec3d5ab0aab106cedc7180c8/operator/doublewrite/metricreporter.go#L86-L88